### PR TITLE
C22 Phoenix - Liubov D.

### PR DESCRIPTION
**Pull request by LiubovDav**

### CODE REVIEW

File: src/components/ChatEntry.jsx [hunk 1]
```diff
@@ -1,20 +1,35 @@
...
 };
 
 ChatEntry.propTypes = {
-  // Fill with correct proptypes
+  id: PropTypes.number.isRequired,
+  sender: PropTypes.string.isRequired,
+  body: PropTypes.string.isRequired,
+  timeStamp: PropTypes.string.isRequired,
+  liked: PropTypes.bool.isRequired,
+  onToggleLiked: PropTypes.func
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:32]: 👍 The `id`, `sender`, `body`, `timeStamp`, and `liked` props are always passed (they're defined explicitly in the data and also provided in the test) so we can (and should) mark them `isRequired`, as you did.

The remaining props were up to you, and the tests don't know about them. As a result, using `isRequired` causes a warning when running any tests that only pass the known props.

To properly mark any other props `isRequired`, we would also need to update the tests to include at least dummy values (such as an empty callback `() => {}` for the like handler) to make the proptype checking happy.

Alternatively, for any props that we leave not required, we should also have logic in our component to not try to use the value if it's undefined.

File: src/components/ChatEntry.jsx [hunk 2]
```diff
@@ -1,20 +1,35 @@
...
-    <div className="chat-entry local">
-      <h2 className="entry-name">Replace with name of sender</h2>
+    <div className={props.sender == 'Vladimir' ? 'chat-entry local' : 'chat-entry remote'} id={props.id}>
+      <h2 className="entry-name">{props.sender}</h2>
       <section className="entry-bubble">
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p>{props.body}</p>
+        <p className="entry-time"><TimeStamp time={props.timeStamp}></TimeStamp></p>
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:19]: Nice use of the supplied `TimeStamp`. All we need to do is pass in the `timeStamp` string from the message data and it takes care of the rest. All we had to do was confirm the name and type of the prop it was expecting (which we could do through its PropTypes) and we're all set!

Note that the `TimeStamp` component doesn't receive children components (it resembles "void" HTML tags such as `input` or `img` rather than tags like `p` or `a` that can have children). For such components, prefer to write them using "self-closing" style (note the `/>` to complete the tag rather than a regular `>`).
```html
<TimeStamp time={props.timeStamp} />
```

File: src/components/ChatLog.jsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import ChatEntry from './ChatEntry';
+import PropTypes from 'prop-types';
+
+const ChatLog = (props) => {
+  const chatEntryComponents = props.entries.map(chatEntry => {
+    return (
+      <ChatEntry
+        key={chatEntry.id}
+        id={chatEntry.id}
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:9]: 👍 The `key` attribute is important for React to be able to detect certain kinds of data changes in an efficient manner. We're also using the `id` for our own `id` prop, so it might feel redundant to pass both, but one is for our logic and one is for React internals (we can't safely access the `key` value in any meaningful way).

File: src/components/ChatLog.jsx [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import ChatEntry from './ChatEntry';
+import PropTypes from 'prop-types';
+
+const ChatLog = (props) => {
+  const chatEntryComponents = props.entries.map(chatEntry => {
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:5]: Nice use of `map` to convert from the message data into `ChatEntry` components. We can perform this mapping storing the result into a variable we use in the JSX result as you did here (components are functions, so we can run JS code as usual before we reach the `return`, and even sometimes have multiple `return` statements with different JSX), we could make a helper function that we call as part of the `return`, or this expression itself could be part of the `return` JSX, which I often like since it helps me see the overall structure of the component, though can make debugging a little more tricky. But any of those approaches will work fine.

File: src/App.jsx [hunk 5]
```diff
@@ -1,14 +1,42 @@
 import './App.css';
+import ChatLog from './components/ChatLog';
+import chatLogDataFromFile from '../src/data/messages.json';
```
Comment (anselrognlie) [inline src/App.jsx:3]: This file is alreasy in `src`, so we don't need to move up a folder and come back down again
```js
import chatLogDataFromFile from './data/messages.json';
```

File: src/App.jsx [hunk 6]
```diff
@@ -1,14 +1,42 @@
...
+import { useState } from 'react';
 
 const App = () => {
+  const [likesCount, setLikesCount] = useState(0);
```
Comment (anselrognlie) [inline src/App.jsx:7]: 👀 It's not necessary to introduce additional state to track the total number of likes, and really we would prefer to avoid doing so. The total number of likes can be calculated from the message data, since each message tracks whether it's liked. The total number of likes is an aggregate property of the message data.

It can be tempting to introduce additional state since it can seem inefficient to iterate over all the message data just to calculate that value. But keep in mind that we have to do this anyway just to render the message data into components, so the additional cost is minimal, and avoids potential problems that could lead to states getting out of sync. For a small application like this, the risk is small, but as apps get more complex, we should look for ways to reduce the state used as much as possible.

File: src/App.jsx [hunk 7]
```diff
@@ -1,14 +1,42 @@
...
+
+  const [chatLogData, setChatLogData] = useState(chatLogDataFromFile);
+
+  const toggleChatEntryLiked = (chatEntryId) => {
```
Comment (anselrognlie) [inline src/App.jsx:11]: 👍 Since our state is defined here, we also need to define our mutating function here so that it can "see" the setter function. All we need to receive is the `id` of the message to toggle, which allows us to locate the message to update, and calculate the next state value. We can pass this all the way down to our `ChatEntry` which handles the click event, passing us the `id` of the message that was clicked.

File: src/App.jsx [hunk 8]
```diff
@@ -1,14 +1,42 @@
...
+    const chatLogDataAfterToggle = chatLogData.map(chatEntry => {
```
Comment (anselrognlie) [inline src/App.jsx:12]: Nice use of map here to both handle making a new list so that React sees the message data has changed, and make new data for the clicked message with its like status toggled.

File: src/App.jsx [hunk 9]
```diff
@@ -1,14 +1,42 @@
...
+      if (chatEntry.id === chatEntryId) {
+        if (chatEntry.liked) {
+          setLikesCount(likesCount - 1);
+        } else {
+          setLikesCount(likesCount + 1);
+        }
```
Comment (anselrognlie) [inline src/App.jsx:18]: If we exclude the additional state to track the total likes, we wouldn't need this logic here. We would need a similar check in whatever logic we use to calculate the total on the next render, but being able to remove the extra state makes moving the logic worth it.

File: src/App.jsx [hunk 10]
```diff
@@ -1,14 +1,42 @@
...
+        return { ...chatEntry, liked: !chatEntry.liked };
+      } else {
+        return chatEntry;
+      }
+    });
+
+    setChatLogData(chatLogDataAfterToggle);
```
Comment (anselrognlie) [inline src/App.jsx:25]: In this case, calculating the next version of the message data using the current state variable and passing that updated version to the state setter shouldn't cause any problems, but we still generally prefer using the callback style of setters. Using that approach, we pass a function to the setter whose parameter will receive the latest state value, and which returns the new value to use for the state.
```js
    setChatLogData(chatData => chatLogData.map(chatEntry => {
      if (chatEntry.id === id) {
        return {...chatEntry, liked: !chatEntry.liked};
      } else {
        return chatEntry;
      };
    }));
```

File: src/components/ChatEntry.jsx [hunk 11]
```diff
@@ -1,20 +1,35 @@
 import './ChatEntry.css';
+import TimeStamp from './TimeStamp.jsx';
+import PropTypes from 'prop-types';
+import { useState } from 'react';
+
+const ChatEntry = (props) => {
+  const [isLiked, setIsLiked] = useState(props.liked);
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:7]: 👀 We don't need (and shouldn't make) an additional piece of state to copy the `liked` prop. The `liked` value will already be managed and passed down by the `App`. It may look like this would update the local state when this happens, but that's not the case. The value passed to `useState` is used only on the first render, so to keep this local state in sync with the global state, the `ChatEntry` has to duplicate the logic that also exists in `App`. This local state violates the idea of lifting state up.

File: src/components/ChatEntry.jsx [hunk 12]
```diff
@@ -1,20 +1,35 @@
...
+
+  const likeButtonClicked = () => {
+    setIsLiked(!props.liked);
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:10]: 👀 This logic is only required because of the local state used for the `liked` value. This duplicates logic also found in `App` and requires this presentation component to have extra information about how to manage the data for the chat. We should remove the additional state, and then we wouldn't need this logic.

File: src/components/ChatEntry.jsx [hunk 13]
```diff
@@ -1,20 +1,35 @@
...
+    props.onToggleLiked(props.id);
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:11]: 👍 Passing the `id` of this message lets the logic defined up in the `App` find the message to update in its data.

File: src/components/ChatEntry.jsx [hunk 14]
```diff
@@ -1,20 +1,35 @@
...
-      <h2 className="entry-name">Replace with name of sender</h2>
+    <div className={props.sender == 'Vladimir' ? 'chat-entry local' : 'chat-entry remote'} id={props.id}>
+      <h2 className="entry-name">{props.sender}</h2>
       <section className="entry-bubble">
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p>{props.body}</p>
+        <p className="entry-time"><TimeStamp time={props.timeStamp}></TimeStamp></p>
+        <button className="like" onClick={likeButtonClicked}>{isLiked ? '❤️' : '🤍'}</button>
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:20]: 👀  We can figure out which emoji to use for the liked status based on the `liked` data for this message. Rather than tracking a local piece of state, we can and should use the prop value here directly. This would let us remove the local state.
```html
        <button className="like" onClick={likeButtonClicked}>{props.liked ? '❤️' : '🤍'}</button>
```

Comment (anselrognlie) [reply src/components/ChatEntry.jsx:20]: 👍 We need a wrapper of some kind rather than calling the received callback through props, since our callback function is expecting a message id as its parameter. If we tried to use it directly as the click event handler, React would end up passing it a clink event, since any function registered as an event handler will always be given the event detail information as its argument.

File: src/components/ChatEntry.jsx [hunk 15]
```diff
@@ -1,20 +1,35 @@
...
+  const likeButtonClicked = () => {
+    setIsLiked(!props.liked);
+    props.onToggleLiked(props.id);
+  };
 
-const ChatEntry = () => {
   return (
-    <div className="chat-entry local">
-      <h2 className="entry-name">Replace with name of sender</h2>
+    <div className={props.sender == 'Vladimir' ? 'chat-entry local' : 'chat-entry remote'} id={props.id}>
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:15]: Nice logic to decide whether to treat a message as local or remote. How could we generalize this so that it didn't need to look explicitly for Vladimir? This project only passes in a single conversation, but ideally, our components should work in other situations.

In the general case, does the `ChatEntry` itself have enough information as it is to "know" which messages are local and which are remote?

**Review by anselrognlie — COMMENTED** [review]
Looks good! Please review my comments as there are a few things I think it's worth taking a closer look at. Let me know if you have any questions.
